pkg/virt-operator/install-strategy/create.go

```diff
@@ -366,185 +713,638 @@ func SyncAll(kv *v1.KubeVirt,
+	imageRegistry := kv.Status.TargetKubeVirtRegistry
+
+	if isServiceClusterIP(service) && isServiceClusterIP(cachedService) && service.Spec.ClusterIP == "" {
+		// Ensure that we always preserve the ClusterIP value from the cached service.
+		// This value is dynamically set when it is equal to "" and can't be mutated once it is set.
+		service.Spec.ClusterIP = cachedService.Spec.ClusterIP
+	}
+
+	// First check if there's anything to do.
+	if objectMatchesVersion(&cachedService.ObjectMeta, imageTag, imageRegistry) {
+		if !isServiceClusterIP(cachedService) {
+			// Type != ClusterIP and the cached service is already up to date. Nothing to do
+			return patchOps, false, nil
+		} else if reflect.DeepEqual(cachedService.Spec, service.Spec) {
+			// spec and annotations are already up to date. Nothing to do
+			return patchOps, false, nil
+		}
+	}
+
+	if !isServiceClusterIP(cachedService) || !isServiceClusterIP(service) {
+		// we're only going to attempt to mutate Type ==ClusterIPs right now because
+		// that's the only logic we have tested.
+
+		// This means both the matching cached service and the new target service must be of
+		// type "ClusterIP" for us to attempt any Patch operation
+		deleteAndReplace = true
+	} else if cachedService.Spec.ClusterIP != service.Spec.ClusterIP {
+
+		// clusterIP is not mutable. A ClusterIP == "" will mean one is dynamically assigned.
+		// It is okay if the cached service has a ClusterIP and the target one is empty. We just
+		// ensure the cached ClusterIP is carried over in the Patch.
+		deleteAndReplace = true
+	}
+
+	if deleteAndReplace {
+		// spec can not be merged or mutated. The only way to update is to replace.
+		return patchOps, deleteAndReplace, nil
+	}
+
+	updateLabels := false
+	mergeSpec := false
+	replaceSpec := false
+
+	// If version labels match, only ensure the spec is completely replaced by the target
+	// once the infrastructure has rolled over. THis ensures we don't clobber a
+	// merged spec during an update.
+	if objectMatchesVersion(&cachedService.ObjectMeta, imageTag, imageRegistry) {
+
+		// if we're rolled over, make sure the spec is replaced.
+		// The spec might not match in the event that items were merged.
+		if infrastructureRolledOver && !reflect.DeepEqual(cachedService.Spec, service.Spec) {
+			replaceSpec = true
+		}
+	} else {
+		// Versions don't match, so ensure labels are updated and that
+		// the specs are merged until the infrastructure rolls over.
+		updateLabels = true
+
+		// If the Specs don't equal each other, decide whether to merge them
+		// or replace based on whether or not the new infrastructure has rolled over.
+		if !reflect.DeepEqual(cachedService.Spec, service.Spec) {
+			if infrastructureRolledOver {
+				replaceSpec = true
+			} else {
+				mergeSpec = true
 			}
-			log.Log.V(2).Infof("rolebinding %v updated", rb.GetName())
+		}
+	}
 
-		} else {
-			log.Log.V(4).Infof("rolebinding %v already exists", rb.GetName())
+	if updateLabels {
```

Comment:
I am not sure but maybe https://github.com/evanphx/json-patch/blob/master/merge.go#L184 can help you by letting the library create the patch for you.

Reply:
yeah, maybe. it's probably a matter of taste, but I like doing a little extra leg work to have complete control over how the patch is generated. If we were doing anything more complicated that overwriting fields, then I'd be more inclined to use a tool like this.